File: src/pronote/url.ts
export const cleanPronoteUrl = (url: string): string => {
  let pronote_url = new URL(url);
  // Clean any unwanted data from URL.
  pronote_url = new URL(`${pronote_url.protocol}//${pronote_url.host}${pronote_url.pathname}`);

  // Clear the last path if we're not main selection menu.
  const paths = pronote_url.pathname.split("/");
  if (paths[paths.length - 1].includes(".html")) {
    paths.pop();
  }

  // Rebuild URL with cleaned paths.
  pronote_url.pathname = paths.join("/");

  // Return rebuilt URL without trailing slash.
  return pronote_url.href.endsWith("/")
    ? pronote_url.href.slice(0, -1)
    : pronote_url.href;
};

File: lib/src/utils/url.dart
Uri? loginPortalUrlToInstanceRootUri (final String url) {
  var uri = Uri.tryParse(url);
  if (uri == null) return null;
  // Clean any unwanted data from URL.
  uri = Uri.parse('${uri.origin}${uri.path}');

  // Clear the last path if we're not main selection menu.
  final paths = uri.path.split("/").where((path) => path.isNotEmpty).toList();
  if (paths.last.endsWith(".html")) {
    paths.removeLast();
  }

  // Rebuild URL with cleaned paths.
  uri = uri.replace(path: paths.join("/"));

  // Return rebuilt URL.
  return uri;
}
